app/upgrades/v20/upgrades.go

```diff
@@ -22,10 +45,86 @@ func CreateUpgradeHandler(
 
 		vm, err := mm.RunMigrations(ctx, configurator, vm)
 		if err != nil {
-			return vm, err
+			return vm, errorsmod.Wrapf(err, "running module migrations")
+		}
+
+		ctx.Logger().Info("Initializing MaxProviderConsensusValidators parameter...")
+		InitializeMaxProviderConsensusParam(ctx, keepers.ProviderKeeper)
+
+		ctx.Logger().Info("Setting MaxValidators parameter...")
+		err = SetMaxValidators(ctx, *keepers.StakingKeeper)
+		if err != nil {
+			return vm, errorsmod.Wrapf(err, "setting MaxValidators during migration")
+		}
+
+		ctx.Logger().Info("Initializing LastProviderConsensusValidatorSet...")
+		err = InitializeLastProviderConsensusValidatorSet(ctx, keepers.ProviderKeeper, *keepers.StakingKeeper)
+		if err != nil {
+			return vm, errorsmod.Wrapf(err, "initializing LastProviderConsensusValSet during migration")
 		}
 
 		ctx.Logger().Info("Upgrade v20 complete")
 		return vm, nil
 	}
 }
+
+// InitializeMaxProviderConsensusParam initializes the MaxProviderConsensusValidators parameter.
+// It is set to 180, which is the current number of validators participating in consensus on the Cosmos Hub.
+// This parameter will be used to govern the number of validators participating in consensus on the Cosmos Hub,
+// and takes over this role from the MaxValidators parameter in the staking module.
+func InitializeMaxProviderConsensusParam(ctx sdk.Context, providerKeeper providerkeeper.Keeper) {
+	params := providerKeeper.GetParams(ctx)
+	params.MaxProviderConsensusValidators = NewMaxProviderConsensusValidators
+	providerKeeper.SetParams(ctx, params)
+}
+
+// SetMaxValidators sets the MaxValidators parameter in the staking module to 200,
+// which is the current number of 180 plus 20.
+// This is done in concert with the introduction of the inactive-validators feature
+// in Interchain Security, after which the number of validators
+// participating in consensus on the Cosmos Hub will be governed by the
+// MaxProviderConsensusValidators parameter in the provider module.
+func SetMaxValidators(ctx sdk.Context, stakingKeeper stakingkeeper.Keeper) error {
+	params, err := stakingKeeper.GetParams(ctx)
+	if err != nil {
+		return err
+	}
+
+	params.MaxValidators = NewMaxValidators
+
+	err = stakingKeeper.SetParams(ctx, params)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// InitializeLastProviderConsensusValidatorSet initializes the last provider consensus validator set
+// by setting it to the first 180 validators from the current validator set of the staking module.
+func InitializeLastProviderConsensusValidatorSet(
+	ctx sdk.Context, providerKeeper providerkeeper.Keeper, stakingKeeper stakingkeeper.Keeper,
+) error {
+	vals, err := stakingKeeper.GetBondedValidatorsByPower(ctx)
+	if err != nil {
+		return err
+	}
+
+	// cut the validator set to the first 180 validators
```

Comment:
Isn't `vals` going to contain exactly 180 validators during the upgrade?
If yes, maybe we should return an error here because something is off?

Reply:
I think it's probably ok like this